### ISSUE TITLE
Pgp encryption

### ISSUE THREAD

**Issue by anon97945** [NekoX-Dev/NekoX#222]
**Is your feature request related to a problem? Please describe.**
Since pgp is implemented in would like to encrypt messages too.

**Describe the solution you'd like**
Maybe add another checkbox for "encrypt message"

**Comment by arichr** [comment]
Devs, may the priority of this feature be raised?

**Comment by arichr** [comment]
@anon97945 , should be the content encrypted too? And one more thing.. GPG clients like Kleopatra on desktop can encrypt contents without signing. Should be data be signed too or leave them as is?

**Comment by anon97945** [comment]
Content in form of files? Would be a to long message if u would try to encrypt a media file. it should be optional to sign? i would love to have it signed.

Or do you mean content by message content...? Yeah, i would like to write e2ee in a group if this would be possible ;) but then you need to hook into the openkeys app and get the reciever key.

**Comment by arichr** [comment]
@anon97945 ,
> Content in form of files? Would be a to long message if u would try to encrypt a media file.
I think there're two possible solutions: 
 * Attach a message as a .bin file. So, it'll look like just you and your interlocutor are exchanging files.
 * Encrypt every part of the message. For example, when you send a document with a caption. Text will be encrypted separately, and then the document will be encrypted and attached to your message.

> Or do you mean content by message content...? 
Yes, I consider text, files, etc. to be the part of message content.

> and get the reciever key.
I don't see any problems with exchanging public keys right in Telegram. But, someone can think it's insecure, so this operation should be optional.

**Comment by anon97945** [comment]
> I think there're two possible solutions:
> 
>     * Attach a message as a .bin file. So, it'll look like just you and your interlocutor are exchanging files.
> 
>     * Encrypt every part of the message. For example, when you send a document with a caption. Text will be encrypted separately, and then the document will be encrypted and attached to your message.
> 
true, using bin files for all the text would work. if it would be well implemented and not feel like 1990 emails. i prefer something seemless.

> 
> Yes, I consider text, files, etc. to be the part of message content.
> 

would be awesome.

> 
> I don't see any problems with exchanging public keys right in Telegram. But, someone can think it's insecure, so this operation should be optional.

i would not like to use telegram, instead use openkeychain or so

**Comment by arichr** [comment]
> i would not like to use telegram, instead use openkeychain or so

Ah. I got it. :ok_hand:

**Comment by FlaNya** [comment]
Regarding encryption with GPG, that would be a good idea. I think the best way is to encrypt text messages and send them encrypted text, but that after the exchange of keys with the interlocutor they are encrypted and decrypted automatically, without breaking the usability of the interface and reading them, and as a label about encryption in the interface next to such encrypted messages make a lock icon for example.
I think it is better to make encryption of files and attachments a separate option, as downloading, processing and decrypting such files, especially large ones, will take some time and space in the phone memory due to caching. Encrypted files can be sent not as text, but as ordinary attachments, the format as above said .bin or standard .gpg

**Comment by arichr** [comment]
@FlaNya ,
> but that after the exchange of keys with the interlocutor

I'm not sure that **everyone** thinks that's secure. I don't doubt about the safety of Telegram, but this exchange should be optional.

> they are encrypted and decrypted automatically

Great idea. I'd want to see this in action.

> especially large ones, will take some time and space in the phone memory due to caching

That can be a problem.. But devs can make checkboxes to disable/enable the encryption per file.

> Encrypted files can be sent not as text, but as ordinary attachments

So, do you want encrypt everything and send as .gpg (or .bin) attachments? Or do we need to separate parts of a message?

**Comment by arm64v8a** [comment]
I don’t use PGP and I think it’s not good to add too many pgp utils to the client.

Maybe other developers are interested in this, so waiting...

**Comment by centerpoint67** [comment]
https://github.com/pgpainless/pgpainless - worth noting

**Comment by MuratovAS** [comment]
I've implemented a basic encryption feature, but it's not good enough to get into the mainline.
It is quite ready for use by advanced users. The finished apk is also in the repository.
I would be glad to have PR. Later I plan to merge with the main branch.
what should be done:
- Ability to select a key for signing (if it is not specified, as well as with a long press), by analogy with the "sign" button.
- Ability to save the message recipient's key for the duration of the chat session (for the lifetime of the chat).
- Send a long message as a file, not text 

https://github.com/MuratovAS/NekoX/tree/pgp-encrypt

**Comment by MartinX3** [comment]
Please add OpenKeyChain support.
https://github.com/open-keychain/open-keychain
